File: crates/shamir-query-builder/src/batch/handle.rs
use shamir_query_types::filter::FilterValue;

use crate::val::{qref, qref_all, IntoFieldPath};

/// A typed reference to a query registered in a [`super::Batch`].
///
/// Returned by `Batch::query`, `Batch::insert`, etc. Its methods produce
/// `FilterValue::QueryRef` values that the engine's batch planner
/// interprets as inter-query dependencies.
#[derive(Debug, Clone)]
pub struct Handle {
    pub(super) alias: String,
}

impl Handle {
    /// The bare alias (without `@` prefix).
    pub fn alias(&self) -> &str {
        &self.alias
    }

    /// Reference a column across all result rows.
    ///
    /// Produces a `$query` path like `"[].field"` (or `"[].a.b"` for
    /// nested fields).
    pub fn column(&self, field: impl IntoFieldPath) -> FilterValue {
        let segments = field.into_field_path();
        let dotted = segments.join(".");
        let path = format!("[].{dotted}");
        qref(&self.alias, path)
    }

    /// Reference a specific result row by index.
    pub fn row(&self, index: usize) -> RowRef {
        RowRef {
            alias: self.alias.clone(),
            index,
        }
    }

    /// Shorthand for `row(0)`.
    pub fn first(&self) -> RowRef {
        self.row(0)
    }

    /// Reference the entire result (no path).
    pub fn all(&self) -> FilterValue {
        qref_all(&self.alias)
    }
}

/// A reference to a specific row of a query result.
#[derive(Debug, Clone)]
pub struct RowRef {
    alias: String,
    index: usize,
}

impl RowRef {
    /// Reference a field on this row.
    ///
    /// Produces a `$query` path like `"[0].field"`.
    pub fn field(&self, field: impl IntoFieldPath) -> FilterValue {
        let segments = field.into_field_path();
        let dotted = segments.join(".");
        let path = format!("[{}].{dotted}", self.index);
        qref(&self.alias, path)
    }

    /// Reference the whole row (no field).
    ///
    /// Produces a `$query` path like `"[0]"`.
    pub fn get(&self) -> FilterValue {
        let path = format!("[{}]", self.index);
        qref(&self.alias, path)
    }
}

File: crates/shamir-client-ts/src/core/builders/handle.ts
/**
 * `Handle` / `RowRef` — typed accessors for `$query` result references inside a
 * batch. Mirrors `crates/shamir-query-builder/src/batch/handle.rs`.
 *
 * These classes build `$query` `FilterValue`s through the existing
 * {@link queryRef} constructor — they do NOT duplicate the wire-path format.
 *
 * `Batch.handle(alias)` returns a `Handle`; chaining on `Batch.add()` is
 * unaffected (it still returns `this`).
 *
 * PLATFORM-AGNOSTIC.
 */

import type { FilterValue } from '../types/filter.js';
import { queryRef } from './filter.js';

/** Normalise a field spec to the dotted-path string used after `[]`. */
function dotted(field: string | string[]): string {
  return typeof field === 'string' ? field : field.join('.');
}

/**
 * Typed handle to a batch query result identified by `alias`.
 *
 * - `.column(field)` → `$query` path `"[].field"` (all rows, one column);
 *   nested paths produce `"[].a.b"`.
 * - `.row(i)` → a {@link RowRef} for row `i`.
 * - `.first()` → shorthand for `.row(0)`.
 * - `.all()` → a bare `$query` ref with no path (the whole result set).
 */
export class Handle {
  constructor(private readonly alias: string) {}

  /** Reference every row's `field` value → `{ $query, path: "[].field" }`. */
  column(field: string | string[]): FilterValue {
    return queryRef(this.alias, `[].${dotted(field)}`);
  }

  /** Reference row `index` → a {@link RowRef}. */
  row(index: number): RowRef {
    return new RowRef(this.alias, index);
  }

  /** Shorthand for `.row(0)`. */
  first(): RowRef {
    return this.row(0);
  }

  /** Reference the whole result set (no path) → `{ $query }`. */
  all(): FilterValue {
    return queryRef(this.alias);
  }
}

/**
 * Typed reference to a single row `index` of a batch query result.
 *
 * - `.field(f)` → `{ $query, path: "[i].field" }`;
 * - `.get()` → `{ $query, path: "[i]" }` (the whole row object).
 */
export class RowRef {
  constructor(
    private readonly alias: string,
    private readonly index: number,
  ) {}

  /** Reference `field` within this row → `{ $query, path: "[i].field" }`. */
  field(field: string | string[]): FilterValue {
    return queryRef(this.alias, `[${this.index}].${dotted(field)}`);
  }

  /** Reference the whole row object → `{ $query, path: "[i]" }`. */
  get(): FilterValue {
    return queryRef(this.alias, `[${this.index}]`);
  }
}
